new_platforms/libstdio/enc/optee/files_optee.c

```diff
@@ -28,6 +30,29 @@ typedef struct _OPTEE_FILE {
 
 int errno = 0;
 
+/* OP-TEE secure storage (RPMB) is not quite a filesystem: there are no
+ * directories and there are no file names. Instead, one stores objects, each
+ * of which is identified by an at most 64-byte long byte sequence. This byte
+ * sequence need not be a sensible string; it is merely treated as a (void *).
+ * Seeing as oe_fopen(...) is used with file paths and that these can be longer
+ * than 64 characters (i.e. bytes), we don't pass the file path down to OP-TEE.
+ * Instead, we hash it first using SHA-512, which conveniently generates a
+ * 64-byte long digest and use that as the object ID.
+ * 
+ * NOTE: This breaks file enumeration.
```

Comment:
Why is file enumeration broken?   Are you fixing it?   There should be a manifest which lists the object IDs in order to make enumeration work.

Reply:
I think this is somewhat of a complicated situation.

On the one hand, there are the OE API's, namely `oe_*`. On the other, there are the Win32-style API's, such as `FindFirstFile` and friends, along with your manifest API's, `AppendToManifest` and family. The former set of API's does not use manifest files while the latter does. This is one side of the coin.

The other side is the way in which files are identified in SGX vs OP-TEE. In SGX, files are stored in the REE FS, so directories and files are used as usual. In OP-TEE's RPMB, there are no directories and no files. Instead, persistent objects are identified by a sequence of up to 64 bytes. This sequence could be a string, or it could be meaningless; the lookup is done with a `memcmp`.

While it is possible to use a string as a sequence of bytes, the problem is the length limitation. In that sense, we could simply raise the limit in OP-TEE. However, if we did that, we would be out of spec. with respect to the GlobalPlatform Internal Core API (table 5-4b: `TEE_OBJECT_ID_MAX_LEN`).

The other route is what I did, which is to take the file path passed into the `oe_*` API's and, when compiling for OP-TEE, use the SHA-512 hash of the path as the object identifier. The side effect of this is, of course, that when enumerating persistent objects, the object IDs are indeed a sequence of meaningless bytes (i.e., the hash), and the original file path is lost. The `oe_*` API's do not yet support file enumeration, but the Win32-style API's do.

Fixing this problem ultimately requires thinking about a filesystem; in this instance, at least a `hash -> filepath` map. Seeing as the OE team are already working on a filesystem spec, I would be doing throwaway work if I try to fix this in any way.

Ultimately, we need their spec. and we need to merge.

Do you think I should do the work to map hashes to their original file paths, or do you think it's best to take a breaking change once the FS API is ready?

Reply:
Thinking a little more, it seems that regardless of what the OE team do in their FS spec, we'll still need to deal with the lack of a proper filesystem in OP-TEE. The issue then is that doing a simple `hash -> filepath` map won't suffice to implement operations like directory enumeration, for example.

Reply:
I'm suggesting that we store in RPMB a manifest blob that contains the set of filepaths for which entries exist in RPMB.  Enumeration can then just read that blob.   Whenever we create a new blob that is supposed to look like a file to the app, we update the manifest blob.   We already do that for SGX in order to protect against deletion attacks.  It's just that we now have a reason to do it for RPMB as well.